Sources/NIOOpenSSL/SSLPrivateKey.swift

```diff
@@ -14,6 +14,95 @@
+/// protocol. However, that causes us an issue, because we need to stuff that callback into an
+/// `OpenSSLPassphraseCallbackManager` in order to create an `Unmanaged` and round-trip the pointer through C code.
+///
+/// That makes `OpenSSLPassphraseCallbackManager` a generic object, and now we're in *real* trouble, becuase
+/// `Unmanaged` requires us to specify the *complete* type of the object we want to unwrap. In this case, we
+/// don't know it, because it's generic!
+///
+/// Our way out is to note that while the class itself is generic, the only function we want to call in the
+/// `globalOpenSSLPassphraseCallback` is not. Thus, rather than try to hold the actual specific `OpenSSLPassphraseManager`,
+/// we can hold it inside a protocol existential instead, so long as that protocol existential gives us the correct
+/// function to call. Hence: `CallbackManagerProtocol`, a private protocol with a single conforming type.
+private protocol CallbackManagerProtocol {
+    func invoke(buffer: UnsafeMutableBufferPointer<CChar>) -> CInt
+}
+
+
+/// This class exists primarily to work around the fact that Swift does not let us stuff
+/// a closure into an `Unmanaged`. Instead, we use this object to keep hold of it.
+final class OpenSSLPassphraseCallbackManager<Bytes: Collection>: CallbackManagerProtocol where Bytes.Element == UInt8 {
+    private let userCallback: OpenSSLPassphraseCallback<Bytes>
+
+    init(userCallback: @escaping OpenSSLPassphraseCallback<Bytes>){
+        // We have to type-erase this.
+        self.userCallback = userCallback
+    }
+
+    func invoke(buffer: UnsafeMutableBufferPointer<CChar>) -> CInt {
+        var count: CInt = 0
+
+        do {
+            try self.userCallback { passphraseBytes in
+                // If we don't have enough space for the passphrase plus NUL, bail out.
+                guard passphraseBytes.count < buffer.count else { return }
+                _ = buffer.initialize(from: passphraseBytes.lazy.map { CChar($0) })
+                count = CInt(passphraseBytes.count)
+
+                // We need to add a NUL terminator, in case the user did not.
+                buffer[Int(passphraseBytes.count)] = 0
+            }
+        } catch {
+            // If we hit an error here, we just need to tolerate it. We'll return zero-length.
+            count = 0
+        }
+
+        return count
+    }
+}
+
+
+/// Our global static OpenSSL passphrase callback. This is used as a thunk to dispatch out to
+/// the user-provided callback.
+func globalOpenSSLPassphraseCallback(buf: UnsafeMutablePointer<CChar>?,
+                                     size: CInt,
+                                     rwflag: CInt,
+                                     u: UnsafeMutableRawPointer?) -> CInt {
+    guard let buf = buf, let u = u else {
+        preconditionFailure("Invalid pointers passed to passphrase callback")
```

Comment:
should we add the values to the string just so it's easier to debug just in case?